src/databricks/sql/auth/authenticators.py

```diff
@@ -0,0 +1,78 @@
+# Databricks CLI
+# Copyright 2022 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"), except
+# that the use of services to which certain application programming
+# interfaces (each, an "API") connect requires that the user first obtain
+# a license for the use of the APIs from Databricks, Inc. ("Databricks"),
+# by creating an account at www.databricks.com and agreeing to either (a)
+# the Community Edition Terms of Service, (b) the Databricks Terms of
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from databricks.sql.auth.oauth import get_tokens, check_and_refresh_access_token
+import base64
+
+
+class Authenticator:
+    def add_auth_token(self, request_headers):
+        pass
+
+
+class AccessTokenAuthenticator(Authenticator):
+    def __init__(self, access_token):
+        self.__authorization_header_value = "Bearer {}".format(access_token)
+
+    def add_auth_token(self, request_headers):
+        request_headers['Authorization'] = self.__authorization_header_value
+
+
+class UserPassAuthenticator(Authenticator):
+    def __init__(self, username, password):
+        auth_credentials = "{username}:{password}".format(username, password).encode("UTF-8")
+        auth_credentials_base64 = base64.standard_b64encode(auth_credentials).decode("UTF-8")
+
+        self.__authorization_header_value = "Basic {}".format(auth_credentials_base64)
+
+    def add_auth_token(self, request_headers):
+        request_headers['Authorization'] = self.__authorization_header_value
+
+
+class OAuthAuthenticator(Authenticator):
+    # TODO: moderakh the refresh_token is only kept in memory. not saved on disk
+    # hence if application restarts the user may need to re-authenticate
+    # I will add support for this outside of the scope of current PR.
+    def __init__(self, hostname, client_id, scopes):
```

Comment:
hostname should not be a property of credentials provider and it's to be taken from configuration.

Reply:
@nfx I am not sure if I understood what you meant in this comment.

`DatabricksOAuthProvider` hides the oauth implementation, so it needs to be aware of hostname or redirect url to initiate oauth protocol. 

in the python sql connector, the end point information is passed as following
```python 
connection = sql.connect(
      server_hostname='moderakh-test.cloud.databricks.com',
      http_path='sql/protocolv1/o/***',
      auth_type="databricks-oauth"
)
```

the implementation passed `server_hostname` down to `DatabricksOAuthProvider` to be used by oauth implementation, there is no independent configuration support.